### PR TITLE
bugfix: buffer bloat and CPU 100% when download large file was filtered by body_filter_by_lua.

### PR DESCRIPTION
**Pull request by zhuizhuhaomeng**
fix issue #1906
I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

### CODE REVIEW

File: src/ngx_http_lua_bodyfilterby.c [hunk 1]
```diff
@@ -272,6 +269,50 @@ ngx_http_lua_body_filter(ngx_http_request_t *r, ngx_chain_t *in)
...
     }
 
+    if (in != NULL) {
+        ll = &ctx->in_bufs;
+        while (*ll != NULL) {
+            ll = &(*ll)->next;
+        }
+
+        for (ln = in; ln != NULL; ln = ln->next) {
+            cl = ngx_alloc_chain_link(r->pool);
```
Comment (agentzh) [inline src/ngx_http_lua_bodyfilterby.c:279]: We should avoid alloc chain (and free chain below). Small memory blocks can never get freed in nginx memory pools. We should recycle them instead.

File: src/ngx_http_lua_bodyfilterby.c [hunk 2]
```diff
@@ -300,8 +344,21 @@ ngx_http_lua_body_filter(ngx_http_request_t *r, ngx_chain_t *in)
...
-    if (in == out) {
-        return ngx_http_next_body_filter(r, in);
+    if (in != out) {
+        /* content of body was replaced in ngx_http_lua_body_filter_param_set
+         * and the buffer was marked as consumed.
+         */
+        for (cl = in; cl != NULL; cl = ln) {
+            ln = cl->next;
+            if (ngx_buf_size(cl->buf) == 0) {
+                ngx_free_chain(r->pool, cl);
```
Comment (agentzh) [inline src/ngx_http_lua_bodyfilterby.c:354]: Ditto.

File: src/ngx_http_lua_bodyfilterby.c [hunk 3]
```diff
@@ -300,8 +344,21 @@ ngx_http_lua_body_filter(ngx_http_request_t *r, ngx_chain_t *in)
...
 
-    if (in == out) {
-        return ngx_http_next_body_filter(r, in);
+    if (in != out) {
+        /* content of body was replaced in ngx_http_lua_body_filter_param_set
+         * and the buffer was marked as consumed.
+         */
+        for (cl = in; cl != NULL; cl = ln) {
+            ln = cl->next;
+            if (ngx_buf_size(cl->buf) == 0) {
```
Comment (agentzh) [inline src/ngx_http_lua_bodyfilterby.c:353]: This does not look right to me. 0-sized can be special bufs indicating flush, eof, and etc.

**Review by agentzh — CHANGES_REQUESTED** [review]

File: src/ngx_http_lua_bodyfilterby.c (hunk 1 above)
Comment (zhuizhuhaomeng) [reply src/ngx_http_lua_bodyfilterby.c:279]: ngx_alloc_chain_link  allocate chain from the chain pool first and ngx_free_chain free the chain to the chain pool.
ngx_chain_t *
ngx_alloc_chain_link(ngx_pool_t *pool)
{
    ngx_chain_t  *cl;

    cl = pool->chain;

    if (cl) {
        pool->chain = cl->next;
        return cl;
    }

    cl = ngx_palloc(pool, sizeof(ngx_chain_t));
    if (cl == NULL) {
        return NULL;
    }

    return cl;
}

#define ngx_free_chain(pool, cl)                                             \
    (cl)->next = (pool)->chain;                                              \
    (pool)->chain = (cl)

File: src/ngx_http_lua_bodyfilterby.c (hunk 2 above)
Comment (zhuizhuhaomeng) [reply src/ngx_http_lua_bodyfilterby.c:354]: see explanation above.

**Comment by waterdeng** [comment]
It works for #1906.

Thank you .

**Review by agentzh — APPROVED** [review]

**Review by agentzh — APPROVED** [review]

File: src/ngx_http_lua_bodyfilterby.c [hunk 4]
```diff
@@ -269,7 +266,50 @@ ngx_http_lua_body_filter(ngx_http_request_t *r, ngx_chain_t *in)
...
         }
 
-        return NGX_OK;
+        in = NULL;
+
+        /* continue to call ngx_http_next_body_filter to process cached data */
+    }
+
+    if (in != NULL) {
+        if (ngx_chain_add_copy(r->pool, &ctx->filter_in_bufs, in) != NGX_OK) {
```
Comment (doujiang24) [inline src/ngx_http_lua_bodyfilterby.c:275]:
```suggestion
    if (in != NULL
        && ngx_chain_add_copy(r->pool, &ctx->filter_in_bufs, in) != NGX_OK)
     {
```

File: src/ngx_http_lua_bodyfilterby.c [hunk 5]
```diff
@@ -269,7 +266,50 @@ ngx_http_lua_body_filter(ngx_http_request_t *r, ngx_chain_t *in)
...
+        }
+
+        out = NULL;
+        ngx_chain_update_chains(r->pool,
+                                &ctx->free_bufs, &ctx->filter_busy_bufs, &out,
+                                (ngx_buf_tag_t) &ngx_http_lua_module);
+        if (rc != NGX_OK) {
+            if (ctx->filter_busy_bufs != NULL
+                && (r->connection->buffered
+                    & (NGX_HTTP_LOWLEVEL_BUFFERED | NGX_LOWLEVEL_BUFFERED))) {
```
Comment (doujiang24) [inline src/ngx_http_lua_bodyfilterby.c:305]:
```suggestion
        if (rc != NGX_OK
             && ctx->filter_busy_bufs != NULL
             && (r->connection->buffered
                 & (NGX_HTTP_LOWLEVEL_BUFFERED | NGX_LOWLEVEL_BUFFERED)))
         {
```

File: src/ngx_http_lua_common.h [hunk 6]
```diff
@@ -548,6 +548,9 @@ typedef struct ngx_http_lua_ctx_s {
     ngx_chain_t             *busy_bufs;
     ngx_chain_t             *free_recv_bufs;
 
+    ngx_chain_t             *filter_in_bufs;  /* for the body filter */
+    ngx_chain_t             *filter_busy_bufs;  /* for the body filter */
```
Comment (doujiang24) [inline src/ngx_http_lua_common.h:552]: can we use union here to save memory?
like:
```
union {
    ngx_chain_t *filter_busy_bufs;
    ngx_chain_t *busy_bufs;
};
```

Maybe `filter_in_bufs` can union with `free_recv_bufs`?

File: src/ngx_http_lua_output.c [hunk 7]
```diff
@@ -244,6 +244,11 @@ ngx_http_lua_ngx_echo(lua_State *L, unsigned newline)
         return 2;
     }
 
+    ngx_log_debug2(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                   "%s has %sbusy bufs",
+                   newline ? "lua say response" : "lua print response",
+                   ctx->busy_bufs != NULL ? "" : "not ");
```
Comment (doujiang24) [inline src/ngx_http_lua_output.c:250]:
```suggestion
                   ctx->busy_bufs != NULL ? "" : "no ");
```

**Review by doujiang24 — APPROVED** [review]

**Pull request merged**